File: ruby/word-count/word_count.rb
class Phrase
  attr_reader :word

  def initialize(word)
    @word = word
  end

  def word_count
    word
      .downcase
      .scan(/\b\w+'?\w*\b/)
      .group_by(&:itself).transform_values(&:count)
  end
end

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    sentence
    |> split_sentence
    |> normalize_case
    |> Enum.reduce(%{}, fn word, acc ->
      Map.update(acc, word, 1, &(&1 + 1))
    end)
  end

  defp split_sentence(sentence) do
    Regex.scan(~r/[[:alnum:]'-]+/u, sentence)
    |> Enum.map(fn [word] -> word end)
  end

  defp normalize_case(words),
    do: Enum.map(words, &String.downcase/1)
end
